test/test_v2/pmi2_utils.c

```diff
@@ -0,0 +1,156 @@
+static inline unsigned char pmi_base64_encsym (unsigned char value) {
+    assert (value < 64);
+
+    if (value < 26) {
+    return 'A' + value;
+    } else if (value < 52) {
+    return 'a' + (value - 26);
+    } else if (value < 62) {
+    return '0' + (value - 52);
+    }
+
+    return (62 == value) ? '+' : '/';
+}
+
+static inline unsigned char pmi_base64_decsym (unsigned char value) {
+    if ('+' == value) {
+    return 62;
+    } else if ('/' == value) {
+    return 63;
+    } else if (' ' == value) {
+    return 64;
+    } else if (value <= '9') {
+    return (value - '0') + 52;
+    } else if (value <= 'Z') {
+    return (value - 'A');
+    } else if (value <= 'z') {
+    return (value - 'a') + 26;
+    }
+
+    return 64;
+}
+
+static inline void pmi_base64_encode_block (const unsigned char in[3], char out[4], int len) {
+    out[0] = pmi_base64_encsym (in[0] >> 2);
+    out[1] = pmi_base64_encsym (((in[0] & 0x03) << 4) | ((in[1] & 0xf0) >> 4));
+    /* Cray PMI doesn't allow = in PMI attributes so pad with spaces */
+    out[2] = 1 < len ? pmi_base64_encsym(((in[1] & 0x0f) << 2) | ((in[2] & 0xc0) >> 6)) : ' ';
+    out[3] = 2 < len ? pmi_base64_encsym(in[2] & 0x3f) : ' ';
+}
+
+static inline int pmi_base64_decode_block (const char in[4], unsigned char out[3]) {
+    char in_dec[4];
+
+    in_dec[0] = pmi_base64_decsym (in[0]);
+    in_dec[1] = pmi_base64_decsym (in[1]);
+    in_dec[2] = pmi_base64_decsym (in[2]);
+    in_dec[3] = pmi_base64_decsym (in[3]);
+
+    out[0] = in_dec[0] << 2 | in_dec[1] >> 4;
+    if (64 == in_dec[2]) {
+        //fprintf(fileptr, "base64_decode_block = %x %x %x\n", out[0], out[1], out[2]);
+        //fflush(fileptr);
+        return 1;
+    }
+
+    out[1] = in_dec[1] << 4 | in_dec[2] >> 2;
+    if (64 == in_dec[3]) {
+        //fprintf(fileptr, "base64_decode_block = %x %x %x\n", out[0], out[1], out[2]);
+        //fflush(fileptr);        
+        return 2;
+    }
+
+    out[2] = ((in_dec[2] << 6) & 0xc0) | in_dec[3];
+    //fprintf(fileptr, "base64_decode_block = %x %x %x\n", out[0], out[1], out[2]);
+    //fflush(fileptr);
+    return 3;
+}
+
+
+/* PMI only supports strings. For now, do a simple base64. */
+char *pmixt_encode(const void *val, size_t vallen)
+{
+    char *outdata, *tmp;
+    size_t i;
+
+    outdata = calloc (((2 + vallen) * 4) / 3 + 2, 1);
+    if (NULL == outdata) {
+    return NULL;
+    }
+
+    for (i = 0, tmp = outdata ; i < vallen ; i += 3, tmp += 4) {
+        pmi_base64_encode_block((unsigned char *) val + i, tmp, vallen - i);
+    }
+
+    tmp[0] = (unsigned char)'\0';
+
+    return outdata;
+}
+
+char *pmixt_decode (const char *data, size_t *retlen)
+{
+    size_t input_len = strlen(data) / 4;
+    unsigned char *ret;
+    int out_len;
+    size_t i;
+    /*fileptr = fopen("client.out", "w");
+
+    fprintf(fileptr, "inside pmixt_decode, data = %s\n", data);
+    fflush(fileptr);
+    fprintf(fileptr, "*retlen = %lu\n", *retlen); */
+
```

Comment:
REMOVE

Reply:
should be fixed (in newly named file, which is base64_enc_dec.c)